extensions/opc/src/opc.cpp

```diff
@@ -0,0 +1,480 @@
+  switch (nd.dataTypeID) {
+    case UA_TYPES_STRING:
+    case UA_TYPES_LOCALIZEDTEXT:
+    case UA_TYPES_BYTESTRING: {
+      UA_String value = *(UA_String * )(nd.var_->data);
+      ret_val = std::string(reinterpret_cast<const char *>(value.data), value.length);
+      break;
+    }
+    case UA_TYPES_BOOLEAN:
+      bool b;
+      memcpy(&b, nd.data.data(), sizeof(bool));
+      ret_val = b ? "True" : "False";
+      break;
+    case UA_TYPES_SBYTE:
+      int8_t i8t;
+      memcpy(&i8t, nd.data.data(), sizeof(i8t));
+      ret_val = std::to_string(i8t);
+      break;
+    case UA_TYPES_BYTE:
+      uint8_t ui8t;
+      memcpy(&ui8t, nd.data.data(), sizeof(ui8t));
+      ret_val = std::to_string(ui8t);
+      break;
+    case UA_TYPES_INT16:
+      int16_t i16t;
+      memcpy(&i16t, nd.data.data(), sizeof(i16t));
+      ret_val = std::to_string(i16t);
+      break;
+    case UA_TYPES_UINT16:
+      uint16_t ui16t;
+      memcpy(&ui16t, nd.data.data(), sizeof(ui16t));
+      ret_val = std::to_string(ui16t);
+      break;
+    case UA_TYPES_INT32:
+      int32_t i32t;
+      memcpy(&i32t, nd.data.data(), sizeof(i32t));
+      ret_val = std::to_string(i32t);
+      break;
+    case UA_TYPES_UINT32:
+      uint32_t ui32t;
+      memcpy(&ui32t, nd.data.data(), sizeof(ui32t));
+      ret_val = std::to_string(ui32t);
+      break;
+    case UA_TYPES_INT64:
+      int64_t i64t;
+      memcpy(&i64t, nd.data.data(), sizeof(i64t));
+      ret_val = std::to_string(i64t);
+      break;
+    case UA_TYPES_UINT64:
+      uint64_t ui64t;
+      memcpy(&ui64t, nd.data.data(), sizeof(ui64t));
+      ret_val = std::to_string(ui64t);
+      break;
+    case UA_TYPES_FLOAT:
+      if(sizeof(float) == 4 && std::numeric_limits<float>::is_iec559){
+        float f;
+        memcpy(&f, nd.data.data(), sizeof(float));
+        ret_val = std::to_string(f);
+      } else {
+        throw Exception(OPC_EXCEPTION, "Float is non-standard on this system, OPC data cannot be extracted!");
+      }
+      break;
+    case UA_TYPES_DOUBLE:
+      if(sizeof(double) == 8 && std::numeric_limits<double>::is_iec559){
+        double d;
+        memcpy(&d, nd.data.data(), sizeof(double));
+        ret_val = std::to_string(d);
+      } else {
+        throw Exception(OPC_EXCEPTION, "Double is non-standard on this system, OPC data cannot be extracted!");
+      }
+      break;
+    case UA_TYPES_DATETIME: {
+      UA_DateTime dt;
+      memcpy(&dt, nd.data.data(), sizeof(UA_DateTime));
+      ret_val = opc::OPCDateTime2String(dt);
+      break;
+    }
+    default:
+      throw Exception(OPC_EXCEPTION, "Data type is not supported ");
+  }
+  return ret_val;
+}
+
+std::string OPCDateTime2String(UA_DateTime raw_date) {
+  UA_DateTimeStruct dts = UA_DateTime_toStruct(raw_date);
+  std::array<char, 100> charBuf;
+
+  snprintf(charBuf.data(), charBuf.size(), "%02hu-%02hu-%02hu %02hu:%02hu:%02hu.%03hu", dts.day, dts.month, dts.year, dts.hour, dts.min, dts.sec, dts.milliSec);
```

Comment:
And I messed this up, the year should be %04hu, which is not an issue, except for dates before 1000 CE, but still.

Reply:
Fixed